src/sentry/api/serializers/models/organization_member/expand/projects.py

```diff
@@ -1,50 +1,73 @@
     def __init__(self, *args: Any, **kwargs: Any) -> None:
-        self.project_ids = set(kwargs.pop("project_ids", []))
+        self.projects = {p.id: p for p in kwargs.pop("projects", [])}
+        self.project_ids = set(self.projects.keys())
         super().__init__(*args, **kwargs)
 
     def get_attrs(
         self, item_list: Sequence[OrganizationMember], user: User, **kwargs: Any
     ) -> MutableMapping[OrganizationMember, MutableMapping[str, Any]]:
-        """
-        Note: For this to be efficient, call
-        `.prefetch_related('teams')`
-        on your queryset before using this serializer
-        """
-
         attrs = super().get_attrs(item_list, user)
 
+        # Get all the OrganizationMemberTeam relations so we can map
+        # the member id to the list of team ids
+        #
+        # Note that we're intentionally only working with `team_id`
+        # to avoid having to fetch the team model as well.
+        member_teams = OrganizationMemberTeam.objects.filter(
+            organizationmember_id__in=[om.id for om in item_list],
+            team_id__in=ProjectTeam.objects.filter(
+                project_id__in=self.project_ids,
+                # make sure to filter only for teams that are visible
+                team__status=TeamStatus.VISIBLE,
+            )
+            .values_list("team_id", flat=True)
+            .distinct(),
+        )
+
+        # The set of team ids, this will be used to filter down the `ProjectTeam` below
         team_ids = set()
-        for org_member in item_list:
-            for team in org_member.teams.all():
-                team_ids.add(team.id)
 
-        projects_by_team = defaultdict(list)
+        # Mapping from member id to team ids they belong to.
+        #
+        # Previously, we were using a `select_related` when fetching the OrganizationMember.
+        # This resulted in django trying to set the attributes for the many to many relation
+        # which was slow.
+        team_ids_by_member_id = defaultdict(list)
+
+        # Be very careful here. We are intentionally only using `team_id` and `organizationmember_id`.
+        # This is to stop django from fetching these models. We don't even want django to do
+        # any kind of prefetching here.
+        for member_team in member_teams:
+            team_ids.add(member_team.team_id)
+            team_ids_by_member_id[member_team.organizationmember_id].append(member_team.team_id)
+
+        # Mapping from team id to projects that belong to the team.
+        #
+        # We require the caller to pass in the list of projects (not just ids to avoid an extra query)
+        # Make sure we only work with `team_id` and not the team object so django doesn't fetching it.
+        projects_by_team_id = defaultdict(list)
 
         for project_team in ProjectTeam.objects.filter(
             project_id__in=self.project_ids,
             team_id__in=team_ids,
-        ).select_related("project"):
```

Comment:
I'm not sure that this `select_related` would cause that much of a performance hit, I think it was just the `prefetch_related`? nbd either way though

Reply:
I don't think this particular `select_related` caused a big performance hit but I removed it since it was unnecessary here.